### ISSUE TITLE
CP1252 encoding on windows

### ISSUE THREAD

**Issue by adament** [rpy2/rpy2#754]
For current (<=4) versions of R on Windows R.dll uses the CP1252 encoding for output in RTerm and LinkDLL CharacterMode. However rpy2 attempts to decode output using the utf-8 encoding irrespective of platform.

**To Reproduce**
In a jupyter notebook I tried the following:

```python
# %%
%load_ext rpy2.ipython

# %%
%%R
sessionInfo()

# %%
%%R
x <- "ěščřžýáíé"
x
```

And got the following output in the second cell:
```
R version 4.0.3 (2020-10-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 17763)

Matrix products: default

locale:
[1] LC_COLLATE=Danish_Denmark.1252  LC_CTYPE=Danish_Denmark.1252   
[3] LC_MONETARY=Danish_Denmark.1252 LC_NUMERIC=C                   
[5] LC_TIME=Danish_Denmark.1252    

attached base packages:
[1] tools     stats     graphics  grDevices utils     datasets  methods  
[8] base     

loaded via a namespace (and not attached):
[1] compiler_4.0.3
```

And an error in the third cell:
```
From cffi callback <function _consolewrite_ex at 0x0000018300FD4288>:
Traceback (most recent call last):
  File "C:\Code\miniconda3\envs\renv\lib\site-packages\rpy2\rinterface_lib\callbacks.py", line 131, in _consolewrite_ex
    s = conversion._cchar_to_str_with_maxlen(buf, maxlen=n)
  File "C:\Code\miniconda3\envs\renv\lib\site-packages\rpy2\rinterface_lib\conversion.py", line 130, in _cchar_to_str_with_maxlen
    s = ffi.string(c, maxlen).decode('utf-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x9a in position 3: invalid start byte
```

**Workaround**
I "fixed" the issue by monkeypatching rpy2 right after loading the rpy2 ipython extension:
```python
%load_ext rpy2.ipython

import rpy2
def _cchar_to_str(c, encoding: str = 'cp1252') -> str:
    # TODO: use isStrinb and installTrChar
    s = rpy2.rinterface_lib.openrlib.ffi.string(c).decode(encoding)
    return s


def _cchar_to_str_with_maxlen(c, maxlen: int) -> str:
    # TODO: use isStrinb and installTrChar
    s = rpy2.rinterface_lib.openrlib.ffi.string(c, maxlen).decode('cp1252')
    return s
rpy2.rinterface_lib.conversion._cchar_to_str = _cchar_to_str
rpy2.rinterface_lib.conversion._cchar_to_str_with_maxlen = _cchar_to_str_with_maxlen
```
Then the output of the third cell became
```
[1] "ešcržýáíé"
```

**Future**
The R project has been working on improving utf-8 support on Windows this year in one project by supporting UTF-8 directly on Windows: https://developer.r-project.org/Blog/public/2020/05/02/utf-8-support-on-windows/ and https://developer.r-project.org/Blog/public/2020/07/30/windows/utf-8-build-of-r-and-cran-packages/
However as explained in the first link the RGui CharacterMode has long had a workaround for a custom utf-8 embedding on Windows. In the beginning of the year it seems that support was added for using this embedding with the LinkDLL CharacterMode https://github.com/wch/r-source/blob/a52d621364ea76cb039c3897b7c910e2e12c2f41/src/include/R_ext/RStartup.h#L96 it might be possible to implement support for decoding this encoding in python, but I have not been able to find any documentation of how it works, I might investigate later if time permits.

**Comment by lgautier** [comment]
Thanks for the report. Do you mean the issue is mostly present with R < 4.0 ?

**Comment by adament** [comment]
No R uses the cp1252 encoding on all officially released versions of Windows so far. They have made an unofficial experimental build of R 4 which uses the UTF-8 encoding, but it is incompatible with a number of compiled modules. In R 4.0 it seems if I read the code correctly that the custom encoding used by the RGui CharacterMode was added to the LinkDLL CharacterMode used by rpy2, however support for this encoding would need to be implemented in rpy2 but would allow full unicode support for rpy2.

**Comment by lgautier** [comment]
Got it. Thanks.

`cp1252` is limited character sets that be all encoded with a single byte, correct?

May the fix would be to drop the use of a default encoding and get from each R string (Char object) the correct encoding.

I do not have a dev Windows system. Can you try the following and tell me what you get? It should work even without your patch.

```r
import rpy2.robjects as ro
ro.r('x <- "ěščřžýáíé" ; y <- "abc"; z <- c(x, y, "百折不撓"); TRUE')
d = dict((name, ro.rinterface.globalenv[name]) for name in 'xyz')
for k, v in d.items():
    print('{}: {}'
          .format(
              k, ', '.join(str(v.get_charsxp(i).encoding) for i in range(len(v)))
          ))
```

On Linux I get:
```
x: CETYPE.CE_UTF8
y: CETYPE.CE_NATIVE
z: CETYPE.CE_UTF8, CETYPE.CE_NATIVE, , CETYPE.CE_UTF8
```

**Comment by OnlyBelter** [comment]
How did you install rpy2 on Windows successfully? Could you write a blog step by step?

**Comment by adament** [comment]
Yes `cp1252` is limited.

I get the following on Windows:
```
x: CETYPE.CE_LATIN1
y: CETYPE.CE_NATIVE
z: CETYPE.CE_LATIN1, CETYPE.CE_NATIVE, CETYPE.CE_NATIVE
```
I am not sure whether or not the latin1 characterisation is correct in this case. When I was first playing around with the problem I attempted to decode the strings using pythons latin1 encoding and some characters got mixed up as far as I recall (I do not think I got any decoding errors) https://i18nqa.com/debug/table-iso8859-1-vs-windows-1252.html describes the difference between cp1252 and latin1. But it might be that I tried to decode strings in the NATIVE encoding using latin1.

**Comment by adament** [comment]
> How did you install rpy2 on Windows successfully? Could you write a blog step by step?

I just installed both r, python and rpy2 using conda and the conda-forge repository. However apart from the encoding problem I think there is a bug in the PATH modification code in rpy2 as I mentioned in the pull-request #752, it seems that the original author has abandonded the pull request, so I am thinking of making a separate pull request for the path issue, it should just be a minor modification to line 123 in suation.py https://github.com/rpy2/rpy2/blob/b37fca2a48553f9a641111b391b77ba57e529217/rpy2/situation.py#L123. For now I work around the problem by running the following code **before** importing rpy2 or using an rpy2 cell in jupyter notebook:
```python
import os
from pathlib import Path
conda_prefix = Path(os.environ['CONDA_PREFIX'])
r_prefix = conda_prefix / 'lib' / 'R'
r_path = str(r_prefix / 'bin' / 'x64')
os.environ['R_HOME'] = str(r_prefix)
paths = os.environ['PATH'].split(';')
if r_path not in paths:
    os.environ['PATH'] = ';'.join([r_path] + paths)
```
note that this assumes you have python and r installed in the **same** conda environment.

**Comment by lgautier** [comment]
> Yes `cp1252` is limited.
> 
> I get the following on Windows:
> 
> ```
> x: CETYPE.CE_LATIN1
> y: CETYPE.CE_NATIVE
> z: CETYPE.CE_LATIN1, CETYPE.CE_NATIVE, CETYPE.CE_NATIVE
> ```
> 
> I am not sure whether or not the latin1 characterisation is correct in this case. When I was first playing around with the problem I attempted to decode the strings using pythons latin1 encoding and some characters got mixed up as far as I recall (I do not think I got any decoding errors) https://i18nqa.com/debug/table-iso8859-1-vs-windows-1252.html describes the difference between cp1252 and latin1. But it might be that I tried to decode strings in the NATIVE encoding using latin1.

The line 
```python
ro.r('x <- "ěščřžýáíé" ; y <- "abc"; z <- c(x, y, "百折不撓"); TRUE')
```
is trying to let evaluate the string as R code the same way it would happen if rpy2 was not involved. Well, sort of. The Python string itself will have to be passed to R using `conversion._str_to_charsxp()`. The Python string is encoded in UTF-8 (function `conversion._str_to_cchar()`) before being passed to R. This part of the chain should be fine because no matter the original encoding of the Python string or your locale UTF-8 is the way things are passed to R.

However, R may decide to encode each string in an array differently. I guess that internally R is using a strategy to minimize the number of bytes used.

My understanding of some of R's own documentation is that strings can only be encoded in Latin1, or in UTF-8
(see https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Character-encoding-issues) and
`CE_NATIVE` will indicate which one of the two is considered the native encoding.

The R man page for `Encoding()` adds interesting information, although potentially confusing one. The confusing one is that  strings can also be tagged with an "unknown" encoding; I don't know what to do in that case so I'll wait until it becomes a problem. The interesting one is:

```txt
     There is some ambiguity as to what is meant by a ‘Latin-1’ locale,
     since some OSes (notably Windows) make use of character positions
     used for control characters in the ISO 8859-1 character set.  How
     such characters are interpreted is system-dependent but as from R
     3.5.0 they are if possible interpreted as per Windows codepage
     1252 (which Microsoft calls ‘Windows Latin 1 (ANSI)’) when
     converting to e.g. UTF-8.
```
Your cp1252 and `Latin1` are both right, somehow. On windows R means cp1252 when saying Latin1.

**Comment by lgautier** [comment]
It turns out that the R-level encoding "unknown" is "other" at the C-level, and means that the encoding is the system's locale. May be. :|
Writing a general patch to handle encodings is a little more involved than I thought. I'll look at it over the next days.

**Comment by lgautier** [comment]
@adament : would you be able to check if PR #757 solves the issue ? Don't hesitate to review the code in the PR. I don't have a Windows system to test this on, so I am guessing a bit (and the final fix required changing quite a few things).

**Comment by adament** [comment]
@lgautier Thank you so much I will definitely, I am looking at it right now - though family obligations might slow my progress.

First I think a typo got into 9b1fa90bd947be90d6bf8a93aae0e5de0f4d2e90 `os.pathjoin` was used but I think it should be [`os.pathsep`](https://docs.python.org/3/library/os.html#os.pathsep).

**Comment by adament** [comment]
So unfortunately the first result is that I get a hard crash when I try to import robjects , but that might be due to me building the module incorrectly (I used the conda-forge recipe since it is their built module that works for me). But I do not know if you can think of any changes to the c code that have happened since 3.3.6 that might cause a crash?

I will try to see if I can narrow down the problem.

**Comment by adament** [comment]
I also get the crash on master but not when building the RELEASE_3_3_6 tag, so I will try to do a git bisect,

**Comment by lgautier** [comment]
Ooops. You are right. Fixed.

**Comment by lgautier** [comment]
Ideally there should be an automated build for Windows as well. I never got
the API mode to successfully build on GitHub Actions + Windows, but the ABI
was working at some point. I did not know that conda had rpy2-3.3.6 builds.

FWIW the GitHub Action file is here:
https://github.com/rpy2/rpy2/blob/master/.github/workflows/pythonpackage.yml
.

**Comment by adament** [comment]
It seems that it was d76393981e63ba4a6cb11abe909bc0807bc7322e which started the crashes. I will try getting a debug version up and running and see if gdb can provide a backtrace or some additional information.

**Comment by lgautier** [comment]
Thanks, that's helpful (although probably worth opening a different issue).
That revision added a poor man's C preprocessor. Its output is probably
generating something that looks valid at build and install time but
ultimately crashes at run time.

By the way, are trying to build with the API mode?

**Comment by adament** [comment]
I use the batch script https://github.com/conda-forge/rpy2-feedstock/blob/master/recipe/bld.bat to build on Windows which sets `compiler=mingw32` and after some dll wrangling just seems to call `pip install . --no-deps -vv`. So if the default is to build with the API mode, I assume that is what I am doing. Is there an easy way for me to identify whether I am trying to build with the API mode?

**Comment by adament** [comment]
I think it became clear from my experiments that I am using the ABI mode.

**Comment by lgautier** [comment]
At build/install time, the default is for rpy2 to try whether a C compiler is available and if the case try to build the API mode (`Tcffi`'s API mode), along with the ABI mode. If building the API mode fails for any reason only the ABI mode will be built/installed. The output in the terminal contains information about what is happening.

At runtime the default is to try loading the API mode first, and it it fails load the ABI mode.

Forcing to only try the ABI or the API mode can be achieved with the environment variable `RPY2_CFFI_MODE` (see https://rpy2.github.io/doc/latest/html/overview.html#easy-install-and-pip).

To get which mode ended being loaded at runtime, you can do:

```python
from rpy2.rinterface_lib import ffi_proxy, openrlib
print('cffi mode: {}'.format(ffi_proxy.get_ffi_mode(openrlib._rinterface_cffi)))
```

**Comment by lgautier** [comment]
Note: The CI has the ABI mode working again on Windows. Tests fail because of an odd errors about timezone handling by a Python module.

The API mode does not build because the setup seems incomplete (`Rinternals.h` is missing).

**Comment by lgautier** [comment]
The fix for CP1252 is in rpy2-3.4.0 (just released) and there is a unit test with coverage the case reported here. I am closing this for now.
